mlx/transforms.cpp

```diff
@@ -193,25 +204,18 @@ void eval(const std::vector<array>& outputs) {
   std::unordered_set<std::uintptr_t> cache;
   std::unordered_map<std::uintptr_t, std::shared_future<void>> deps;
 
-  std::set<std::uintptr_t> output_primitives;
-  for (auto& arr : outputs) {
-    if (!arr.is_evaled()) {
-      output_primitives.insert(arr.primitive_id());
-    }
-  }
+  auto synchronizer = array(
+      {},
+      bool_,
+      std::make_unique<Synchronizer>(default_stream(default_device())),
```

Comment:
It might make sense to choose the stream in a better way. This could in theory still be slow a problem if you eval a bunch of outputs mostly on a non-default stream.

🤔 could try to compute the most common output stream to avoid that..

Another alternative I considered is to not use a primitive at all, simply put a synchronizing task in each actual output stream at the very end. I went with this because it was a lot simpler to implement, but the former might be more bullet proof.